packages/event-handler/src/appsync-graphql/Router.ts

```diff
@@ -0,0 +1,169 @@
+   *
+   * // Register a Query resolver
+   * app.resolver(async (payload) => {
+   *   // your business logic here
+   *   return payload;
+   * }, {
+   *   fieldName: 'getPost'
+   * });
+   *
+   * // Register a Mutation resolver
+   * app.resolver(async (payload) => {
+   *   // your business logic here
+   *   return payload;
+   * }, {
+   *   fieldName: 'createPost',
+   *   typeName: 'Mutation'
+   * });
+   *
+   * export const handler = async (event, context) =>
+   *   app.resolve(event, context);
+   * ```
+   *
+   * You can also specify the type of the arguments using a generic type parameter:
+   *
+   * @example
+   * ```ts
+   * import { AppSyncGraphQLResolver } from '@aws-lambda-powertools/event-handler/appsync-graphql';
+   *
+   * const app = new AppSyncGraphQLResolver();
+   *
+   * app.resolver<{ postId: string }>(async ({ postId }) => {
+   *   // postId is now typed as string
+   *   return { id: postId };
+   * }, {
+   *   fieldName: 'getPost'
+   * });
+   *
+   * export const handler = async (event, context) =>
+   *   app.resolve(event, context);
+   * ```
+   *
+   * As a decorator:
+   *
+   * @example
+   * ```ts
+   * import { AppSyncGraphQLResolver } from '@aws-lambda-powertools/event-handler/appsync-graphql';
+   *
+   * const app = new AppSyncGraphQLResolver();
+   *
+   * class Lambda {
+   *   @app.resolver({ fieldName: 'getPost' })
+   *   async handleGetPost(payload) {
+   *     // your business logic here
+   *     return payload;
+   *   }
+   *
+   *   async handler(event, context) {
+   *     return app.resolve(event, context);
+   *   }
+   * }
+   *
+   * const lambda = new Lambda();
+   * export const handler = lambda.handler.bind(lambda);
+   * ```
+   *
+   * @param handler - The handler function to be called when the event is received.
+   * @param options - Route options including the required fieldName and optional typeName.
+   * @param options.fieldName - The name of the field to register the handler for.
+   * @param options.typeName - The name of the GraphQL type to use for the resolver (defaults to 'Query').
+   */
+  public resolver<TParams extends Record<string, unknown>>(
+    handler: ResolverHandler<TParams>,
+    options: GraphQlRouteOptions
+  ): void;
+  public resolver(options: GraphQlRouteOptions): MethodDecorator;
+  public resolver<TParams extends Record<string, unknown>>(
+    handler: ResolverHandler<TParams> | GraphQlRouteOptions,
+    options?: GraphQlRouteOptions
+  ): MethodDecorator | undefined {
+    if (typeof handler === 'function') {
+      const resolverOptions = options as GraphQlRouteOptions;
+      const { typeName = 'Query', fieldName } = resolverOptions;
+
+      this.resolverRegistry.register({
+        fieldName,
+        handler: handler as ResolverHandler,
+        typeName,
+      });
+
+      return;
+    }
+
+    const resolverOptions = handler;
+    return (_target, _propertyKey, descriptor: PropertyDescriptor) => {
```

Comment:
The decorator implementation will likely have to change to reflect this fix: https://github.com/aws-powertools/powertools-lambda-typescript/pull/3974.